nntrainer/tensor/manager.cpp

```diff
@@ -207,24 +197,67 @@ void Manager::initialize() {
       };
     };
 
-    allocate_weight = get_allocfunc(total_weight_size, weight_mmaped_memory);
-
-    size_t grad_size =
-      enable_gradient_memory_opt ? max_grad_size : total_grad_size;
-    allocate_grad = get_allocfunc(grad_size, grad_mmaped_memory);
-
-  } else {
+    if (is_weight) {
+      /** For weights */
+      allocate_func = get_allocfunc(total_weight_size, weight_mmaped_memory);
+    } else {
+      /** for gradients */
+      size_t grad_size =
+        enable_gradient_memory_opt ? max_grad_size : total_grad_size;
+      allocate_func = get_allocfunc(grad_size, grad_mmaped_memory);
+    }
+  } else if (!is_weight) {
+    /** only for gradients */
     if (max_grad_size > 0 && enable_gradient_memory_opt) {
       std::shared_ptr<float> window(new float[max_grad_size],
                                     std::default_delete<float[]>());
 
-      allocate_grad = [window](const TensorDim &dim, size_t offset) {
+      allocate_func = [window](const TensorDim &dim, size_t offset) {
         return Tensor::Map(window, dim, offset);
       };
     }
   }
 
+  return allocate_func;
+}
+
+/**
+ * @brief Allocate and initialize the weight variable
+ */
+void Manager::initializeWeights() {
+  if (total_weight_size == 0) {
+    ml_logw("Nothing done on initialize because there is no weight registered");
+    return;
+  }
+
+  AllocFunc allocate_weight = getAllocFunc(true);
+
   size_t weight_offset = 0;
+
+  for (auto &l_w : weights) {
+    for (auto &w : l_w) {
+      Weight &weight = w.get();
+      auto dim = weight.getDim();
+      Tensor weight_prealloc = allocate_weight(dim, weight_offset);
+      Tensor grad_prealloc = Tensor();
+
+      weight_offset += dim.getDataLen();
+      weight.initialize(weight_prealloc, Tensor(), false);
```

Comment:
how about separate initializing gradient as well (thus making `weight::initialize` to `weight::initializeWeight`), instead caller having the flexibility to call weights and gradients if needed?

Reply:
Added separate methods.